src/mdo/model.vm

```diff
@@ -145,55 +145,39 @@ public class ${class.name}
       * @see Builder#build()
       */
-    ${class.name}(
-    #if ( $class == $root )
-        String namespaceUri,
-        String modelEncoding,
-    #end
-    #foreach ( $field in $allFields )
-      #set ( $sep = "#if(${locationTracking}||$field!=${allFields[${allFields.size()} - 1]}),#end" )
-      #set ( $type = ${types.getOrDefault($field,${types.getOrDefault($field.type,$field.type)})} )
-      #if ( $type.startsWith("List<") )
-        #set ( $type = ${type.replace('List<','Collection<')} )
-      #end
-        $type $field.name${sep}
-    #end
-    #if ( $locationTracking )
-        Map<Object, InputLocation> locations,
-        InputLocation importedFrom
-    #end
-    ) {
+    protected ${class.name}(Builder builder) {
     #if ( $class.superClass )
-        super(
-      #foreach ( $field in $inheritedFields )
-        #set ( $sep = "#if(${locationTracking}||$field!=${inheritedFields[${inheritedFields.size()} - 1]}),#end" )
-            ${field.name}${sep}
-      #end
-      #if ( $locationTracking )
-            locations,
-            importedFrom
-      #end
-        );
+        super(builder);
     #end
     #if ( $class == $root )
-        this.namespaceUri = namespaceUri;
-        this.modelEncoding = modelEncoding;
+        this.namespaceUri = builder.namespaceUri != null ? builder.namespaceUri : (builder.base != null ? builder.base.namespaceUri : null);
+        this.modelEncoding = builder.modelEncoding != null ? builder.modelEncoding : (builder.base != null ? builder.base.modelEncoding :  "UTF-8");
     #end
     #foreach ( $field in $class.getFields($version) )
       #if ( $field.type == "java.util.List" || $field.type == "java.util.Properties" || $field.type == "java.util.Map" )
-        this.${field.name} = ImmutableCollections.copy(${field.name});
+        this.${field.name} = ImmutableCollections.copy(builder.${field.name} != null ? builder.${field.name} : (builder.base != null ? builder.base.${field.name} : null));
       #else
-        this.${field.name} = ${field.name};
+        #if ( $field.type == "boolean" || $field.type == "int" )
+        this.${field.name} = builder.${field.name} != null ? builder.${field.name} : (builder.base != null ? builder.base.${field.name} : ${field.defaultValue});
+        #else
+        this.${field.name} = builder.${field.name} != null ? builder.${field.name} : (builder.base != null ? builder.base.${field.name} : null);
+        #end
       #end
     #end
     #if ( $locationTracking )
+        Map<Object, InputLocation> newlocs = builder.locations != null ? builder.locations : Collections.emptyMap();
+        Map<Object, InputLocation> oldlocs = builder.base != null && builder.base.locations != null ? builder.base.locations : Collections.emptyMap();
       #if ( ! $class.superClass )
-        this.locations = ImmutableCollections.copy(locations);
-        this.importedFrom = importedFrom;
+        this.locations = new HashMap<>();
```

Comment:
Fields should be immutable.  Maybe the generated should look like:
```
        this.locations = Map.of(
                "", newlocs.containsKey("") ? newlocs.get("") : oldlocs.get(""),
                "modules", newlocs.containsKey("modules") ? newlocs.get("modules") : oldlocs.get("modules"),
                "distributionManagement", newlocs.containsKey("distributionManagement") ? newlocs.get("distributionManagement") : oldlocs.get("distributionManagement"),
                "properties", newlocs.containsKey("properties") ? newlocs.get("properties") : oldlocs.get("properties"),
                "dependencyManagement", newlocs.containsKey("dependencyManagement") ? newlocs.get("dependencyManagement") : oldlocs.get("dependencyManagement"),
                "dependencies", newlocs.containsKey("dependencies") ? newlocs.get("dependencies") : oldlocs.get("dependencies"),
                "repositories", newlocs.containsKey("repositories") ? newlocs.get("repositories") : oldlocs.get("repositories"),
                "pluginRepositories", newlocs.containsKey("pluginRepositories") ? newlocs.get("pluginRepositories") : oldlocs.get("pluginRepositories"),
                "reporting", newlocs.containsKey("reporting") ? newlocs.get("reporting") : oldlocs.get("reporting"));
```

Reply:
This is orthogonal to this PR. The same logic from https://github.com/apache/maven/pull/1660/files#diff-ad6fc545933fb055ede5c7cf4ca4f05c63de54967e3253e4244b00adcaa7dcc9L503-L511 just has been moved to the constructor.

Reply:
Well, no.  The constructor was getting a possibly mutable map, but it was doing an immutable copy before assigning it to the field.

Reply:
I really dislike the knowledge that is being spread across the `Builder#build()` method and the actual constructor.  But the shallow-copy would be really a performance loss during the initial transformation of the model.

I think it would be much easier to simply change the visibility of constructors to `protected`, I think this would be enough to allow extension from outside the package, as the generated code already allows extension (as we have `Model` extending `ModelBase` and other similar classes).

Reply:
> I think it would be much easier to simply change the visibility of constructors to protected

This would require duplicating all of the functionality from the super-classes `Builder.build()` method! This is exactly what I want to prevent with this approach.

The `locations` map contains `null` values therefore cannot leverage `Map.ofEntries`.

Reply:
> > I think it would be much easier to simply change the visibility of constructors to protected
> 
> This would require duplicating all of the functionality from the super-classes `Builder.build()` method! This is exactly what I want to prevent with this approach.

If you find a better way that keep the current features... but the objects have to be immutable.
Honestly, this is a very rare case, so duplicating a few lines of code may not be so dramatic.
And as I said, the model is now supposed to evolve, rather than be hacked externally (with an outdated XSD + model which do not reflect the reality, etc...)

Reply:
Done now in https://github.com/apache/maven/pull/1660/commits/8ce4cdedd2040f9a97b2cc6a586b67b4aae749e6. Please check again.